File: LeetCode/40组合总和Ⅱ.cpp
class Solution {
public:
    void backtracking(vector<int> &candidates, int target, int sum, int ind, vector<bool> &used) {
        // if (ind == candidates.size()) return ;
        if (target == sum) {
            ans.emplace_back(temp);
            return ;
        }
        for (int i = ind; i < candidates.size() && target - candidates[i] >= sum; i++) {
            if (i > 0 &&  candidates[i] == candidates[i - 1] && used[i - 1] == false) continue;
            sum += candidates[i];
            temp.push_back(candidates[i]);
            used[i] = true;
            backtracking(candidates, target, sum, i + 1, used);
            used[i] = false;
            sum -= candidates[i];
            temp.pop_back();
        }
        return ;
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<bool> used(candidates.size(), false);
        sort(candidates.begin(), candidates.end());
        backtracking(candidates, target, 0, 0, used);
        return ans;
    }

private :
    vector<vector<int>> ans;
    vector<int> temp;
};

File: LeetCode/40组合总和Ⅱ.go
var ans [][]int
var used []bool
var nums []int

func combinationSum2(candidates []int, target int) [][]int {
	ans = [][]int{}
	used = make([]bool, len(candidates))
	nums = []int{}
	sort.Ints(candidates)
	backtracking(0, 0, candidates, target)
	return ans
}

func backtracking(ind int, sum int, candidates []int, target int) {
	if sum == target {
		temp := make([]int, len(nums))
		copy(temp, nums)
		ans = append(ans, temp)
		return
	}
	for i := ind; i < len(candidates) && sum+candidates[i] <= target; i++ {
		if i > 0 && candidates[i] == candidates[i-1] && !used[i-1] {
			continue
		}
		used[i] = true
		sum += candidates[i]
		nums = append(nums, candidates[i])
		backtracking(i+1, sum, candidates, target)
		nums = nums[:len(nums)-1]
		sum -= candidates[i]
		used[i] = false
	}
}